ls1intum/Artemis
# `Programming exercises`: Enable static code analysis for Xcode

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


This PR introduces Static Code Analysis for iOS programming exercises (Programming Language: Swift, Project Type: Xcode). It is done with swiftlint, like for plain Swift exercises. But in the case of iOS programming exercises fastlane is used to configure the SCA task on the build agents build plan. 
It is necessary to give the instructor the possibility to configure the static code analysis in the exercise creation process. This means that there is a checkbox shown to enable the SCA and if it is ticked a text field is shown to specify the penalty, exactly as it is done for other programming languages and project types. 
The build plan needs to have have a final task called for the SCA. It should contain the creation of the necessary target directory, the configuration necessary for fastlane as well as the fastlane sca command to run the static code analysis (shown in video 1). 
The build plan should also specify an artifact called swiftlint with the result of the SCA (shown in video 2). 
The results displayed on Artemis not only shows the amount of passed test cases but also the amount of static code analysis issues. This is shown on the exercise overview page (screenshot 1 & 2). In the online code editor (for the instructor) the SCA issue is reported in the corresponding line (screenshot 3). 

### Description

I adapted the UI in the exercise creation process to show the checkbox and afterwards the text field. 
I adapted some internal SCA verification for Xcode.
I added a final script task to the build plan for iOS programming exercises to execute the static code analysis. 
I added a new lane for executing swiftlint and saving the result in a predefined file to the existing fastlane configuration. 

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- TS1 is required!

1. Log in to Artemis as Instructor
2. Navigate to Course Administration
3. Create a new programming exercise
4. Select Swift and Xcode
5. See the SCA checkbox and tick it 
6. See text field and enter value 
7. Finish exercise creation
8. Check build plans (see videos 1 & 2)
9. Check results (screenshots 1-3)
10. Open the online code editor as instructor (click 'edit in editor'-button)
11. Select any .swift file and enter more than one empty line after a { and submit the code. 
12. Wait for the build plan to run 
13. Check that a new issue occurs in the result saying that there are too many empty lines. 

If you want to test the student side regarding the results of the SCS (which will not differ from the instructors side) you need a Mac with a Xcode installation. You need to start the exercise as a student, clone the repository, open it in your local Xcode installation enter one or more SCA issues (see step 11) and push the changes. Afterwards you need to wait for the build plans execution. Last but not least you can see the results as in step 13.

### Review Progress





#### Code Review
#### Manual Tests

### Screenshots


Screenshot 1:
<img width="435" alt="SCA Result_Exercise Overview" src="https://user-images.githubusercontent.com/82283581/151280957-0d1183fc-8c69-45d7-a48a-13956ea937d1.png">

Screenshot 2:
<img width="952" alt="SCA_Issue" src="https://user-images.githubusercontent.com/82283581/151280986-80625a79-2e48-4bc0-9f6c-a0d1c807bd2a.png">

Screenshot 3:
<img width="451" alt="SCA_Issue_Code Editor" src="https://user-images.githubusercontent.com/82283581/151281054-593ec579-5f8d-4744-a4c8-f074f1df9cb9.png">

Video 1:
https://user-images.githubusercontent.com/82283581/151281075-4a0f7650-8c34-4e9c-ae1d-5a396d65b421.mov

Video 2:
https://user-images.githubusercontent.com/82283581/151281292-cdc17ae4-6536-4e97-bdde-cd6b887950a1.mov
